src/section/section_map.rs

```diff
@@ -0,0 +1,461 @@
+        {
+            pair
+        } else {
+            return 0;
+        };
+
+        if let Some(sibling_index) = self.knowledge.get(&prefix.sibling()) {
+            // The sibling section might not have processed the split yet, so it might still be in
+            // `target`'s location. Because of that, we need to return index that would be trusted
+            // by them too.
+            index.min(*sibling_index)
+        } else {
+            index
+        }
+    }
+
+    /// Provide a start index of a SectionProofSlice that proves the given signature to the given
+    /// destination location.
+    /// If `node_knowledge_override` is `Some`, it is used when calculating proof for
+    /// `DstLocation::Node` instead of the stored knowledge. Has no effect for other location types.
+    pub fn knowledge_index(
+        &self,
+        target: &DstLocation,
+        node_knowledge_override: Option<u64>,
+    ) -> u64 {
+        match (target, node_knowledge_override) {
+            (DstLocation::Node(_), Some(knowledge)) => knowledge,
+            _ => self.proving_index(target),
+        }
+    }
+
+    /// Updates the entry in `knowledge` for `prefix` to the `version`; if a split
+    /// occurred in the meantime, the versions for sections covering the rest of the address space
+    /// are initialised to the old version that was stored for their common ancestor
+    pub fn update_knowledge(&mut self, prefix: Prefix<XorName>, version: u64) {
+        trace!(
+            "attempts to update knowledge of our elders_info with version {:?} for \
+             prefix {:?} ",
+            version,
+            prefix
+        );
+
+        if let Some((&old_pfx, &old_version)) = self
+            .knowledge
+            .iter()
+            .find(|(pfx, _)| pfx.is_compatible(&prefix))
+        {
+            if old_version >= version || old_pfx.is_extension_of(&prefix) {
+                // Do not overwrite newer version or prefix extensions
+                return;
+            }
+
+            let _ = self.knowledge.remove(&old_pfx);
+
+            trace!(
+                "    from {:?}/{:?} to {:?}/{:?}",
+                old_pfx,
+                old_version,
+                prefix,
+                version
+            );
+
+            let old_pfx_sibling = old_pfx.sibling();
+            let mut current_pfx = prefix.sibling();
+            while !self.knowledge.contains_key(&current_pfx) && current_pfx != old_pfx_sibling {
+                let _ = self.knowledge.insert(current_pfx, old_version);
+                current_pfx = current_pfx.popped().sibling();
+            }
+        }
+        let _ = self.knowledge.insert(prefix, version);
+    }
+
+    /// Compute an estimate of the size of the network from the size of our routing table.
+    ///
+    /// Return (estimate, exact), with exact = true iff we have the whole network in our
+    /// routing table.
+    pub fn network_size_estimate(&self) -> (u64, bool) {
```

Comment:
Ah this thing, I know this is not new code, but just pointing out that it is a little bit confusing to call this network size when it is just counting Elders.

Reply:
To be network size perhaps we need elder count * 60 as min and elder count * 120 as max estimate? To compromise elder count * our_current_section_size  and assume relatively even distribution.

Reply:
I'd just rename the method to something like `network_elder_count_estimate` and let the upper layers do the total node count estimation if they want to.